[export] Add aot_export 1/N

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #101490

This PR adds aot_export_module as the lowering path from torch.level graph to aten graph. Some known limitations that need to be addressed in the follow up PRs:
1. Store param/buffer data in ExportedProgram
2. Fully support torch.cond with params/buffers 
3. Making sure no duplicated ExportMetaData entry 
4. This API will break Executorch if used on PyE, we will figure out a plan internally. 


cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

tugsbayasgalan committed:
[WIP] add aot_export

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] add aot_export"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] add aot_export"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] add aot_export"




cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] add aot_export"




cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] add aot_export"




cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] add aot_export"




cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] add aot_export"




cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

tugsbayasgalan commented on test/export/test_export.py:
```diff
@@ -530,8 +530,9 @@ def fn_ddo(x):
         def foo(x):
-            x.add_(4)
-            y = x.view(x.shape)
+            z = x + 4
```
We don't allow input mutations in forward.

tugsbayasgalan commented on torch/_export/__init__.py:
```diff
@@ -106,34 +110,90 @@ def _export(
+
+            fake_mode = detect_fake_mode(fake_inps)
+
+            fake_args = pytree.tree_map_only(torch.Tensor, fake_mode.from_tensor, args)
```
Probably not needed as we are making every placeholder entry to have fake tensor but just to be safe.

tugsbayasgalan commented on test/export/test_passes.py:
```diff
@@ -34,42 +34,14 @@ def test_replace_broken_ops(self) -> None:
         self.assertEqual(count_after, 0)
-        self.assertTrue(torch.allclose(gm(x), f(x)))
-
-    def test_const_prop_pass(self) -> None:
```
There is nothing to const prop anymore

tugsbayasgalan committed:
Update on "[WIP] add aot_export"


This PR adds aot_export_module as the lowering path from torch.level graph to aten graph. Some known limitations that need to be addressed in the follow up PRs:
1. Store param/buffer data in ExportedProgram
2. Fully support torch.cond with params/buffers 
3. Making sure no duplicated ExportMetaData entry 


cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ydwu4 commented on torch/_export/__init__.py:
```diff
@@ -106,34 +110,90 @@ def _export(
+
+            gm_torch_level.recompile()
+            gm, graph_signature = aot_export_module(gm_torch_level, fake_args, decompositions=DECOMP_TABLE, trace_joint=False)
+            mutation_data = MutationData(
```
why we want to have a new data structure MutationData instead of using graph_signature directly?

tugsbayasgalan replied:
There are some fields in GraphSignature such as input_spec and out_spec which is different from what we actually serialize in ExportMetaData. To not confuse users, i decided to filter out things we don't need rn

ydwu4 replied:
Ok. MutationData looks a bit unintuitive though. We could consider naming it as ExportSignature? 

Or does it make sense to override in_spec and out_spec in GraphSignature? Since the backward_signature is optional and may be used in the future. The benefit is less name to remember hahah.

tugsbayasgalan replied:
Yeah I agree. Let's sync today to decide on what to do with them.

avikchaudhuri replied:
What was the resolution of this meeting?

tugsbayasgalan replied:
Basically Angela's ExportedProgram diff was the result of this sync.

ydwu4 commented on torch/_export/graph_module.py:
```diff
@@ -256,26 +279,32 @@ def make_export_graph_module(
     attach_export_graph_metadata(gm, meta)
+    # TODO figure out where to save them
+
+    gm.parameters = params if params is not None else OrderedDict()
```
QQ: what's the difference between params and buffers vs mutation_data.parameters , mutation_data.buffers?

tugsbayasgalan replied:
mutation_data.params and mutation_data.buffers are the names (str) of the params and buffers, not actual data

tugsbayasgalan committed:
Update on "[export] Add aot_export 1/N"


This PR adds aot_export_module as the lowering path from torch.level graph to aten graph. Some known limitations that need to be addressed in the follow up PRs:
1. Store param/buffer data in ExportedProgram
2. Fully support torch.cond with params/buffers 
3. Making sure no duplicated ExportMetaData entry 
4. This API will break Executorch if used on PyE, we will figure out a plan internally. 


cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

avikchaudhuri reviewed (approved):
Accepting to unblock. Rebase and retest. @tugsbayasgalan says this can land without https://github.com/pytorch/pytorch/pull/102310

tugsbayasgalan committed:
Update on "[export] Add aot_export 1/N"


This PR adds aot_export_module as the lowering path from torch.level graph to aten graph. Some known limitations that need to be addressed in the follow up PRs:
1. Store param/buffer data in ExportedProgram
2. Fully support torch.cond with params/buffers 
3. Making sure no duplicated ExportMetaData entry 
4. This API will break Executorch if used on PyE, we will figure out a plan internally. 


cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[export] Add aot_export 1/N"


This PR adds aot_export_module as the lowering path from torch.level graph to aten graph. Some known limitations that need to be addressed in the follow up PRs:
1. Store param/buffer data in ExportedProgram
2. Fully support torch.cond with params/buffers 
3. Making sure no duplicated ExportMetaData entry 
4. This API will break Executorch if used on PyE, we will figure out a plan internally. 


cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[export] Add aot_export 1/N"


This PR adds aot_export_module as the lowering path from torch.level graph to aten graph. Some known limitations that need to be addressed in the follow up PRs:
1. Store param/buffer data in ExportedProgram
2. Fully support torch.cond with params/buffers 
3. Making sure no duplicated ExportMetaData entry 
4. This API will break Executorch if used on PyE, we will figure out a plan internally. 


cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[export] Add aot_export 1/N"


This PR adds aot_export_module as the lowering path from torch.level graph to aten graph. Some known limitations that need to be addressed in the follow up PRs:
1. Store param/buffer data in ExportedProgram
2. Fully support torch.cond with params/buffers 
3. Making sure no duplicated ExportMetaData entry 
4. This API will break Executorch if used on PyE, we will figure out a plan internally. 


cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[export] Add aot_export 1/N"


This PR adds aot_export_module as the lowering path from torch.level graph to aten graph. Some known limitations that need to be addressed in the follow up PRs:
1. Store param/buffer data in ExportedProgram
2. Fully support torch.cond with params/buffers 
3. Making sure no duplicated ExportMetaData entry 
4. This API will break Executorch if used on PyE, we will figure out a plan internally. 


cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]
